### ISSUE TITLE
Release v2.0.0. of Ceph CSI driver.

### ISSUE THREAD

**Issue by humblec** [ceph/ceph-csi#557]
- [x] **Define release process**
  - Release doc would be a good start
- [x] **Define release deadlines**
  - Code Freeze : 15-Dec-2019
  - RC : 20-Dec-2019
  - GA : 29-Dec-2019

- [x] **Define/list roadmap for v2.0.0**
  - Proposed contents for v.2.0.0 release
  - [x] Resize support for CephFS and RBD plugin
     Issues: 
      #101 
      #660 
  
  - [x] Snapshot/Cloning support for CephFS and RBD.
     #394
     #258
     #701
     #411
     #702

  - [x]Topology Aware Provisioning.
     #440
  - [ ] Native Client ( Go client) Support-
    #449
  - [x] Locking support for RBD volumes which are RWO
    #578 
  - [x] RBD NBD support.
    #667
  - [x] Manage RBD image mirroring via StorageClass parameter
    #676
  
- [x] **Release Team**:
 @Madhu-1
 @humblec 
 
  - The release blockers (issues/PRs) are labeled with `release-2.0.0`.
- [x] **Document features and announce the release.**

**Comment by dillaman** [comment]
Tracking here as well that RBD locking support will not land in the v2.0 due to missing external dependencies.

**Comment by ajarr** [comment]
@humblec @Madhu-1, I don't see the clone of FS subvolume snapshots feature being implemented in Ceph master, and backported to nautilus in time for CSI to be use by Nov 5. 

Yes, we already have `fs subvolume snapshot create/rm` commands. Hence, we can easily support CephFS CSI volume snapshots. Is there a ceph-csi issue for RBD/CephFS snapshot?

The `fs subvolume resize`command  is close to being done in Ceph. With this command you can easily implement resize feature for CSI v2.0.0

The following correctness/performance issues need to be fixed for v2.0.0
https://github.com/ceph/ceph-csi/issues/539
https://github.com/ceph/ceph-csi/issues/360

**Comment by ajarr** [comment]
@humblec @Madhu-1 ^^
Can you clarify? I don't see FS clone getting implemented in time for v2.0.0 dev freeze. So can we remove it?

Let's stick to FS `resize` and `snapshot` for now?

**Comment by Madhu-1** [comment]
moving FS clone out of v2.0.0 release as it is not getting implemented in ceph before cephcsi v2.0.0

**Comment by ajarr** [comment]
> Had a discussion with @humblec regarding cephfs cloning requirement, as this is required feature for CSI v2.0.0 release, and as code freeze date is the concern for implementing the cephfs cloning functionality moved the dev freeze date to `20 Nov`
> 
> cloning covers below 2 things
> 
> * creating a PVC from the snapshot
> * creating a PVC from existing PVC
> 

Looking at the CSI documentation, cloning covers only **creating a PVC from an existing PVC**,
https://kubernetes.io/docs/concepts/storage/volume-pvc-datasource/
https://kubernetes.io/blog/2019/06/21/introducing-volume-cloning-alpha-for-kubernetes/
_"While cloning is similar in behavior to creating a snapshot of a volume, then creating a volume from the snapshot, a clone operation is more streamlined and is more efficient for many backend devices."_

**creating a PVC from a snapshot**  seems to be the `Restore Volume from Snapshot` feature
https://kubernetes.io/docs/concepts/storage/persistent-volumes/#volume-snapshot-and-restore-volume-from-snapshot-support

Wouldn't it be better to split the https://github.com/ceph/ceph-csi/issues/411 into two feature requests as follows?
* CephFS: Restore Volume from Snapshot
* CephFS: Volume cloning

@joscollin please follow up on the above.

Regarding **creating a PVC from a snapshot**,
@vshankar will start working on the Ceph side to allow creating FS subvolumes from snapshots. You can track the progress here,
https://tracker.ceph.com/issues/24880

No plans yet on **creating a PVC from existing PVC**. Maybe it will just end up being a volume snapshot followed by creating a volume from snapshot?

> @ajarr is this viable? if more time is required we can move the code freeze date according, please let me know your thoughts

Not sure yet. @joscollin will later update on the time required to implement `creating FS subvolumes from snapshots` feature in Ceph.

**Comment by Madhu-1** [comment]
> > Had a discussion with @humblec regarding cephfs cloning requirement, as this is required feature for CSI v2.0.0 release, and as code freeze date is the concern for implementing the cephfs cloning functionality moved the dev freeze date to `20 Nov`
> > cloning covers below 2 things
> > 
> > * creating a PVC from the snapshot
> > * creating a PVC from existing PVC
> 
> Looking at the CSI documentation, cloning covers only **creating a PVC from an existing PVC**,
> https://kubernetes.io/docs/concepts/storage/volume-pvc-datasource/
> https://kubernetes.io/blog/2019/06/21/introducing-volume-cloning-alpha-for-kubernetes/
> _"While cloning is similar in behavior to creating a snapshot of a volume, then creating a volume from the snapshot, a clone operation is more streamlined and is more efficient for many backend devices."_
> 
> **creating a PVC from a snapshot** seems to be the `Restore Volume from Snapshot` feature
> https://kubernetes.io/docs/concepts/storage/persistent-volumes/#volume-snapshot-and-restore-volume-from-snapshot-support
> 
> Wouldn't it be better to split the #411 into two feature requests as follows?
> 
> * CephFS: Restore Volume from Snapshot
> * CephFS: Volume cloning
> 
> @joscollin please follow up on the above.
> 
> Regarding **creating a PVC from a snapshot**,
> @vshankar will start working on the Ceph side to allow creating FS subvolumes from snapshots. You can track the progress here,
> https://tracker.ceph.com/issues/24880
> 
> No plans yet on **creating a PVC from existing PVC**. Maybe it will just end up being a volume snapshot followed by creating a volume from snapshot?
> 

yes you are correct if we same logic is backed by single cephfs command to create a volume from the volume will be more helpful here

> > @ajarr is this viable? if more time is required we can move the code freeze date according, please let me know your thoughts
> 
> Not sure yet. @joscollin will later update on the time required to implement `creating FS subvolumes from snapshots` feature in Ceph.

@ajarr Thanks for the update.

**Comment by ajarr** [comment]
> Wouldn't it be better to split the #411 into two feature requests as follows?
> 
> * CephFS: Restore Volume from Snapshot
> * CephFS: Volume cloning
> 

Done. It should add more clarity.

**Comment by Madhu-1** [comment]
as cephfs cloning cannot be completed in ceph core before v2.0.0 release, moving it out

CC @ajarr @humblec

**Comment by D1abloRUS** [comment]
@humblec When is this release expected?

**Comment by humblec** [comment]
> @humblec When is this release expected?

@D1abloRUS planning to release at end of this year, however there are some concerns about some of the feature completion which gives us 2 choice atm, that said, either push some of the incomplete features out of `2.0.0` and release `2.1.0` with those or delay the release and include maximum features in this release. We are waiting for couple more weeks to take the call and it should also help us to validate the state in a better way.  I believe you are specifically interested in `resize` functionality , is that the case ? is it something you can use it from `canary` image of master?

**Comment by D1abloRUS** [comment]
@humblec We can pick it up from the master, in any case the request https://github.com/ceph/ceph-csi/pull/331 is still open

**Comment by humblec** [comment]
> @humblec We can pick it up from the master, in any case the request #331 is still open

Its merged last week. You should be able to consume it from master. :+1:

**Comment by humblec** [comment]
[Release call update]

**Feature completed:**

-[x]  CephFS resize
-[x]  RBD resize
-[x]  Encryption support on rbd volumes
-[x]  RBD nbd support -> Available : **Known issue:**  the client mount will be disturbed ( same as fuse deamon)  if node plugin is restarted.


**Waiting for completion:**

-[x] Multi arch support of CSI drivers
-[x] Native go client support  

Deferred from `v2.0.0` to `2.1.0`
-[x]  Topology Aware provisioning
-[x]  Snapshot/Clone of RBD/cephfs support.

**Comment by nixpanic** [comment]
Moving out the native go-ceph implementation for now. Expect it to be ready for release-2.1.0. Performance improvement is currently minimal, as there are still a few calls to the `rbd` executable while creating volumes.

go-ceph will also mature more, and will be better tested by the time 2.1.0 is ready.

**Comment by humblec** [comment]
[Release call update]

Out of pending 2 features, 
- [x] Multi arch support of CSI drivers
- [x] Native go client support

Native go client support has been deferred from the release 2.0.0.
Multi arch support has been merged/completed!

We are at final  stage of building `2.0.0`. It should be available at any time now :)

**Comment by humblec** [comment]
Waiting for  #https://github.com/ceph/ceph-csi/pull/764 to be merged. It should be the last PR for release `2.0.0`

**Comment by ouzklcn** [comment]
Is it possible for you to also consider #785 for `2.0.0` after #764 is merged?

**Comment by humblec** [comment]
> Is it possible for you to also consider #785 for `2.0.0` after #764 is merged?

Sure, let me check.

**Comment by humblec** [comment]
> > Is it possible for you to also consider #785 for `2.0.0` after #764 is merged?
> 
> Sure, let me check.

ACK.   Tagged for `release-2.0.0`

**Comment by humblec** [comment]
All the required PRs are merged.. We are at final stage of `2.0.0` release and build!! Thanks all !

**Comment by Madhu-1** [comment]
closing this as ceph-csi v2.0.0 release  is out https://github.com/ceph/ceph-csi/releases/tag/v2.0.0

**Comment by humblec** [comment]
Thanks to Ceph CSI community for our next major release (`v2.0.0`) !! Many features, bug fixes, documentation updates are part of this release !!  It was a massive help from our community to roll out this big release .. Lets march toward our next big release **v2.1.1** !!! 

@vasyl-purchel @cyb70289 @wilmardo @hswong3i @nixpanic @sophalHong @woohhan @Madhu-1 :+1:  :-1:
